packages/storage/src/adapters/S3Storage/S3Storage.ts

```diff
@@ -0,0 +1,188 @@
+import type { Tag } from '@aws-sdk/client-s3'
+import {
+  GetObjectCommand,
+  DeleteObjectCommand,
+  S3Client,
+} from '@aws-sdk/client-s3'
+import { Upload } from '@aws-sdk/lib-storage'
+import { getSignedUrl } from '@aws-sdk/s3-request-presigner'
+import mime from 'mime-types'
+
+import type { SaveOptionsOverride } from '../BaseStorageAdapter.js'
+import { BaseStorageAdapter } from '../BaseStorageAdapter.js'
+
+export type S3StorageOptions = {
+  // The base directory within the S3 bucket where files will be stored
+  baseDir: string
+  // The name of the S3 bucket to use for storage
+  bucket: string
+  // The AWS region where the S3 bucket is located
+  region: string
+  // The public part of your AWS access key pair, used to identify your AWS account or IAM user
+  accessKeyId: string
+  // The private part of your AWS access key pair, used to sign AWS API requests (keep this secret!)
+  secretAccessKey: string
+  // Optional: Custom endpoint URL for S3-compatible storage services (e.g., Fly/Tigris or MinIO)
+  endpoint?: string
+  // Optional: Cache control header value. Defaults to 1 week.
+  cacheControl?: string
+  // Optional: Tags to apply to the object
+  tags?: Tag[]
+  /**
+   * OptionThe size of the concurrent queue manager to upload parts in parallel. Set to 1 for synchronous uploading of parts. Note that the uploader will buffer at most queueSize * partSize bytes into memory at any given time.
+   * default: 4
+   */
+  queueSize?: number
+  /**
+   * Optional: The size in bytes for each individual part to be uploaded. Adjust the part size to ensure the number of parts does not exceed maxTotalParts.
+   * See 5mb is the minimum allowed part size.
+   */
+  partSize?: number
+  // Optional: Log progress of the upload
+  showProgress?: boolean
+}
+export class S3Storage
+  extends BaseStorageAdapter
+  implements BaseStorageAdapter
```

Comment:
I don't think I've seen this before where you both extend and implement the same base class. What does that actually mean? What does it do?

Reply:
Not sure but again I followed the pattern established that the File and Memory adapters use in current main https://github.com/redwoodjs/redwood/blob/e384c39a9c201264d1b411149896e1afe6772fb0/packages/storage/src/adapters/FileSystemStorage/FileSystemStorage.ts#L12

Reply:
Now that I think about it the base adapter defines the common read etc but some also implement a remove and sign. So make implements makes sure the common methods are implemented and then extend says we have some new custom methods like sign or remove but still consider the class to be like the base adapter? Thus the extend?